File: src/main/kotlin/com/atsuishio/superbwarfare/perk/damage/MagnificentHowl.kt
package com.atsuishio.superbwarfare.perk.damage

import com.atsuishio.superbwarfare.data.gun.DefaultGunData
import com.atsuishio.superbwarfare.data.gun.GunData
import com.atsuishio.superbwarfare.perk.Perk
import com.atsuishio.superbwarfare.perk.PerkInstance
import com.atsuishio.superbwarfare.tools.DamageTypeTool
import net.minecraft.world.damagesource.DamageSource
import net.minecraft.world.entity.Entity

object MagnificentHowl : Perk("magnificent_howl", Type.DAMAGE) {
    override fun computeProperties(
        data: GunData,
        rawData: DefaultGunData
    ): DefaultGunData {
        val tag = data.perk.getTag(this) ?: return super.computeProperties(data, rawData)
        if (tag.getInt("MagnificentHowlDamageCount") > 0) {
            rawData.damage *= 1.5
        }
        return super.computeProperties(data, rawData)
    }

    override fun onKill(
        data: GunData,
        instance: PerkInstance,
        target: Entity,
        source: DamageSource
    ) {
        val tag = data.perk.getTag(this) ?: return
        if (DamageTypeTool.isHeadshotDamage(source)) {
            tag.putInt(
                "MagnificentHowlCount",
                (tag.getInt("MagnificentHowlCount") + 1 + instance.level / 5).coerceAtMost(9 + instance.level)
            )
        }
    }

    override fun preReload(
        data: GunData,
        instance: PerkInstance,
        entity: Entity?
    ) {
        val tag = data.perk.getTag(this) ?: return
        tag.putInt("MagnificentHowlDamageCount", tag.getInt("MagnificentHowlCount"))
        tag.remove("MagnificentHowlCount")
    }

    override fun onHurtEntity(
        damage: Float,
        data: GunData,
        instance: PerkInstance,
        target: Entity,
        source: DamageSource
    ) {
        val tag = data.perk.getTag(this) ?: return
        if (tag.getInt("MagnificentHowlDamageCount") > 0) {
            data.perk.reduceCooldown(this, "MagnificentHowlDamageCount")
        }
    }
}

File: src/main/resources/data/superbwarfare/sbw/perks/magnificent_howl.js
// MAGNIFICENT_HOWL: 爆头击杀累积层数，换弹后获得伤害加成
function modifyProperty(pmc, level, perkTag, gunDataProxy) {
    if (!perkTag) return
    if (perkTag.getInt("MagnificentHowlDamageCount") > 0) {
        pmc.mul("Damage", 1.5)
    }
}

function onKill(perkTag, level, gunData, targetProxy, sourceProxy) {
    if (!perkTag) return
    if (sourceProxy.isHeadshotDamage()) {
        var count = perkTag.getInt("MagnificentHowlCount")
        var newCount = count + 1 + Math.floor(level / 5)
        var maxCount = 9 + level
        perkTag.putInt("MagnificentHowlCount", Math.min(newCount, maxCount))
    }
}

function preReload(perkTag, level, gunData, entityProxy) {
    if (!perkTag) return
    perkTag.putInt("MagnificentHowlDamageCount", perkTag.getInt("MagnificentHowlCount"))
    perkTag.remove("MagnificentHowlCount")
}

function onHurtEntity(damage, perkTag, level, gunData, targetProxy, sourceProxy) {
    if (!perkTag) return
    if (perkTag.getInt("MagnificentHowlDamageCount") > 0) {
        perkTag.reduceCooldown("MagnificentHowlDamageCount")
    }
}
